bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v3/ImageImpl.java

```diff
@@ -0,0 +1,107 @@
+
+import java.util.Set;
+import javax.annotation.PostConstruct;
+import javax.inject.Inject;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.sling.api.SlingHttpServletRequest;
+import org.apache.sling.api.resource.ModifiableValueMap;
+import org.apache.sling.api.resource.PersistenceException;
+import org.apache.sling.api.resource.ResourceResolver;
+import org.apache.sling.models.annotations.Exporter;
+import org.apache.sling.models.annotations.Model;
+import org.apache.sling.models.annotations.injectorspecific.ScriptVariable;
+import org.apache.sling.models.annotations.injectorspecific.SlingObject;
+import org.apache.sling.settings.SlingSettingsService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import com.day.cq.wcm.api.Page;
+import com.adobe.cq.export.json.ComponentExporter;
+import com.adobe.cq.export.json.ExporterConstants;
+import com.adobe.cq.wcm.core.components.internal.servlets.AdaptiveImageServlet;
+import com.adobe.cq.wcm.core.components.models.Image;
+import com.adobe.cq.wcm.core.components.models.Teaser;
+import com.day.cq.commons.Externalizer;
+import com.day.cq.wcm.api.components.Component;
+
+@Model(adaptables = SlingHttpServletRequest.class, adapters = { Image.class,
+        ComponentExporter.class }, resourceType = ImageImpl.RESOURCE_TYPE)
+@Exporter(name = ExporterConstants.SLING_MODEL_EXPORTER_NAME, extensions = ExporterConstants.SLING_MODEL_EXTENSION)
+public class ImageImpl extends com.adobe.cq.wcm.core.components.internal.models.v2.ImageImpl implements Image {
+
+    public static final String RESOURCE_TYPE = "core/wcm/components/image/v3/image";
+    private static final Logger LOGGER = LoggerFactory.getLogger(ImageImpl.class);
+    public final static String OBJECT_ID = "s_objectID=";
+    public final static String APOSTROPHE = "'";
+    public final static String SEMICOLON = ";";
+
+    private String compHashCode = StringUtils.EMPTY;
+    private String hashCodeValue = StringUtils.EMPTY;
+    private StringBuilder linkTrackingCode;
+    private boolean trackingEnabled = false;
+    private Set<String> runmode;
+
+    @ScriptVariable
+    protected Page currentPage;
+
+    @ScriptVariable
+    private Component component;
+
+    @SlingObject
+    private ResourceResolver resourceResolver;
+
+    @Inject
+    private SlingSettingsService settings;
+
+    @PostConstruct
+    protected void initModel() {
+        compHashCode = properties.get(Image.PN_TRACKING_OBJECT_ID, compHashCode);
+        trackingEnabled = properties.get(Image.PN_TRACKING_ENABLED, trackingEnabled);
+        hashCodeValue = String.valueOf(Math.abs(resource.getPath().hashCode() - 1));
+        runmode = settings.getRunModes();
+        if (runmode.contains(Externalizer.AUTHOR)) {
+            populateObjectId();
+        }
+        super.initModel();
+
+    }
+
+    public void populateObjectId() {
+        ModifiableValueMap map = resource.adaptTo(ModifiableValueMap.class);
```

Comment:
Can't the object Id be computed on read? This write will be triggered during the invocation of a GET (idempotent) request. It also requires the page to be opened by a user with write permissions to it.

This seems to also break if an author copy/pastes a component - the tracking code will be the same since its persisted in the copied resource.

Im not clear from the Jira issue if/when this object Id is ok to change over the "life" of the component. (If its moved to a different layout container; should it have the same objectId?)